### DISCUSSION TITLE
grouped selections

### DISCUSSION THREAD

**Discussion by mwotton** [TimefoldAI/timefold-solver#876] [General]
I've got something very much like the Employee Scheduling problem in the quickstart. One of the wrinkles is that there's a hard constraint that certain shifts be done by the same person -  my initial attempt at modelling it was correct in a sense, but it generally didn't find a solution that didn't break some hard constraint. I added a soft constraint that penalised partial pattern matches by the number of shifts that were missed, as well as the hard constraint that they be completely filled, and that seemed to help it gradient-climb its way to a better solution, but I can't help feeling that I'm hammering screws here. Is there a better design pattern to use here? Should the Patterns be planning variables rather than just data?

**Comment by triceo** [comment]
This is likely a problem with generic moves.

Let's assume you have Shifts S1, S2, S3. 
S1 is assigned to Ann.
S2 is assigned to Beth.
S3 is assigned to Carl.
What will the solver do?

It will take one shift, and reassign it. Say it takes S2, and sets it to Ann. S1 and S2 are now on the same person, but S3 still isn't - the hard constraint is still broken! And it will be, until the solver stumbles upon another move which finally sets S3 to Ann.
This is statistically unlikely, and therefore the solution quality will suffer.

You may find some luck with enabling pillar moves.
If that doesn't help, you'll have to write a custom move altogether - one which reassigns multiple shifts to the same person all at once.

**Reply by mwotton** [reply]
yeah, this is why i set a second soft constraint that would prefer Ann having S1&S2 to the original situation - the hard constraint is broken, but the soft constraint is slightly better. that creates some sort of gradient, no?

(It certainly did a lot better but I'll try the pillar move idea. Thanks!)

**Reply by triceo** [reply]
You are right that it is likely to perform better that way, but you may still want to provide moves that can easily eliminate hard constraints. (While at the same time not breaking others.) It's not a problem to include the existing generic moves as well, but introducing these improved moves to the mix will help the solver overall.

**Reply by mwotton** [reply]
hrm. I tried out some pillar moves, but it looks like the initial first-fit construction heuristic phase got a -1hard/6soft  solution that the localsearch step after that wasn't able to improve on very much. how odd - maybe i'd be better off trying to give the initial heuristic a hint that it ought to try to assign all shifts that have this pattern requirement first via FIRST_FIT_DECREASING with a planning difficulty implementation?

**Reply by triceo** [reply]
I'm not entirely sure if what you want to achieve is doable by an iterative incremental approach.
CH has the same issue as LS - it tries one thing after another, it has no idea there are dependencies between entities.

(This is a gap in Timefold Solver's modelling capabilities. There _ought to be_ a way to tell the solver that some things only move together, or not at all.)

A custom phase which pre-assigns the shifts to employees could work better. And then for LS, a custom move which moves multiple things at once.

**Reply by mwotton** [reply]
Thanks, I'll chew on that.
